### ISSUE TITLE
Knex is unable to locate migration files in Node

### ISSUE THREAD

**Issue by uninstallit** [knex/knex#5635]
# Environment

Knex version: 2.4.2
Database + version: pg 8.7.3
OS: Node 18.x

Select applicable template from below.

 @lorefnon 

# Bug

1. Explain what kind of behavior you are getting and how you think it should do

Tried to use Knex to deploy migration scripts from an AWS Lambda function on Node 18 - but Knex is unable to locate migration files from /tmp. 

I verified that files do exist in the tmp folder - and I was expecting Knex to be able to locate the files given proper configuration.  It is worth noting that the Lambda was packaged with Webpack with target="node" , and I am using Knex api to load config and call latest.

2. Error message
Sorry I lost my logs but it was quite clear that it could not locate the files

As a workaround, I was able to read files via third party lib and use a custom source as shown in the Knex documentation; however, this is not ideal. 

So, in summary,  I don't know how Knex is reading the files under the hood - but `fs` can see them but unfortunately` Knex` is having trouble.

**Comment by AdamSGit** [comment]
Can you provide a minimal reproduction environment for this ? Also, does it work if ran outside of an AWS lambda function ?

**Comment by uninstallit** [comment]
Not sure how to provide an environment - but an easier way to replicate would be to use a local docker image from node base. In the image copy a migration file in a `/tmp` folder, copy the main file in `usr/src/main.js`. 

ex. main
```
...
knex.migrate.latest([{directory: "/tmp"}])
...
```

Then run container

**Comment by mercmobily** [comment]
This response comes too late. I know... It will hopefully be helpful to people finding this via search engines.

Thank you for the details. The filesystem and bundling constraints in AWS Lambda commonly cause dynamic file loading to fail unless paths are set explicitly and bundlers are configured correctly.

The filesystem migration source resolves each configured directory relative to `process.cwd()` and lists files at that resolved path; see `lib/migrations/migrate/sources/fs-migrations.js:12` where `const absoluteDir = path.resolve(process.cwd(), configDir)` and then `readdir(absoluteDir)` are used. This means:

- If you provide a relative `migrations.directory`, it must be relative to `process.cwd()` in the Lambda runtime (typically `/var/task`).
- If you provide an absolute path (e.g. `/tmp/migrations`), it will be used as-is, which is the safest approach in Lambda.

Suggested fixes:

- Use an absolute path for migrations in your Knex config, for example `directory: '/tmp/migrations'`, and ensure the files are present there at runtime (copy them during cold start if needed).
- Ensure bundlers (Webpack/Esbuild) do not try to inline or relocate migration files. Preserve `__dirname`/`__filename` and mark migration assets as external or copy them separately.
- Alternatively, avoid the filesystem by providing a programmatic migration source via `migrations: { getMigrations, getMigrationName, getMigration }`, which removes file I/O in Lambda.

Lambda specifics to keep in mind:

- Only `/tmp` is writable; deployment package and layers are read‑only.
- `process.cwd()` may not be where your migrations live; prefer absolute paths.
